text/web-servers/0003-zero-config.md

```diff
@@ -0,0 +1,75 @@
+# Support "zero-configuration" build processes
+
+## Summary
+The NGINX and HTTPD buildpacks within the Web Servers language-family should
+provide a "zero-configuration" option whereby they generate a very simple
+configuration so that the buildpack user does not need to.
+
+## Motivation
+
+Many users have very simple static file deployments. In these cases, requiring
+them to include a bunch of NGINX or HTTPD configuration boilerplate alongside
+their static files can be a burden. Instead, we should allow them to indicate
+that they have some files to serve in the simplest possible way.
+
+## Detailed Explanation
+
+The NGINX and HTTPS buildpacks will implement a common interface through an
+environment variable that will allow them to detect and build when a
+configuration file is not present in the application workspace. Specifically,
+when the `BP_WEB_SERVER` environment variable is set to either `nginx` or
+`httpd` then the respective buildpack will detect and ensure that a simple
+configuration file is included in the built image such that it can serve very
+basic static file applications.
+
+The configuration files will make assumptions about certain options, like the
+location of the files to be served. However, in some cases, the buildpacks
+could choose to allow buildpack users to configure some commonly-specified
+options (e.g. `BP_WEB_SERVER_ROOT=/workspace/my-build-directory`). This RFC
+will not enumerate these options.
+
+## Rationale and Alternatives
+
+### Why not Staticfile?
+This proposal sounds a lot like what the existing Staticfile buildpack is
+already doing, allowing users to indicate that they would like a simple
+configuration for the web server of their choice generated at build time to
+serve their static files.
+
+In the simplest cases, this is as easy as including the following
+`buildpack.yml` file in their codebase:
+
+```yaml
+staticfile:
+  nginx: {}
+```
+
+However, many users find themselves needing more than just the simplest
+configuration. They want support for SSI and HSTS, Basic Auth, and enforcement
+of HTTPS connections. This has created a slippery slope in the Staticfile
+buildpack where it is slowly recreating each of the configuration options
+already available in a proper web server configuration file with a more limited
+API. In most cases, users working with the Staticfile buildpack would likely be
+better off using the NGINX or HTTPD buildpacks directly. They'd have better
+control over their deployment configuration and the complete set of
+configuration options provided by that server.
+
+This also doesn't take into account that the current implementation is
+completely built upon a concept that we've been trying to remove from all of
+the buildpacks in Paketo, the `buildpack.yml` file. We could convert the
+buildpack to read its configuration from a `Staticfile` file like what is done
+in the Cloud Foundry buildpack, but this doesn't do anything to resolve the
+"slippery slope" concern.
```

Comment:
People like having the configuration generated for them. It's convenient.

When you get to the point where you're trying to do something that's more advanced than what is exposed through staticfile config, people are reluctant to switch to generating their own config because they have to start from scratch.

Not sure if there is interest, but if there's a way to have the buildpack export its configuration, or possibly we just document where the configuration lives & how someone could copy it out of the container, I think that would help users transition and in turn help to reduce the desire for the more complicated features and the escape hatches (i.e. please include random config file for me).

Reply:
+1 to documenting how someone could look at the generated config living in the buildpack.

I'd also advocate documenting/creating sample apps that show how to achieve the use-cases you named above, Dan (push state for Java client side apps, basic auth, force HTTPS) with nginx and httpd configuration, rather than adding these as features to a no-config build.

Reply:
I really think we should try to push for some level of config generation beyond the standard static file host scenario. 

At the end of the day, it's what separates using buildpacks from `docker run --name some-nginx -v /some/content:/usr/share/nginx/html:ro -d nginx` (standard static file hosting) or `docker run --name my-custom-nginx-container -v /host/path/nginx.conf:/etc/nginx/nginx.conf:ro -d nginx` (providing a total custom nginx config). The same can be done with the httpd image.

IMHO, config generation is the value-add feature that separates buildpacks from using the stock nginx / httpd images for this use case. I agree that we need reasonable limits set on what we will/won't do for config generation and to keep the buildpacks manageable. I just feel like there's a middle ground here, where we can add some more configuration generation options, but keep them locked into specific, common use cases/scenarios.

Reply:
I've outlined in this RFC that we should document how a user my "eject" out of this zero-config setup into using the `nginx` or `httpd` buildpack. We would document how users could retrieve the generated configuration file from their built image.